multipaz-models/src/commonMain/kotlin/org/multipaz/models/provisioning/ProvisioningModel.kt

```diff
@@ -0,0 +1,353 @@
+                            )
+                        }
+                        KeyBindingInfo.OpenidProofOfPossession(jwtList)
+                    }
+                    // TODO: proof of possession
+                    else -> throw IllegalStateException()
+                }
+            }
+
+            mutableState.emit(RequestingCredentials)
+            val credentials = provisioningClient.obtainCredentials(keyInfo)
+
+            if (credentialMetadata.keyBindingType == KeyBindingType.Keyless) {
+                if (credentials.size != 1) {
+                    throw IllegalStateException("Only a single keyless credential must be issued")
+                }
+                KeylessSdJwtVcCredential.create(
+                    document,
+                    null,
+                    CREDENTIAL_DOMAIN_SD_JWT_VC_KEYLESS,
+                    (format as CredentialFormat.SdJwt).vct
+                )
+            } else {
+                for ((credentialData, pendingCredential) in credentials.zip(pendingCredentials)) {
+                    // TODO: remove validity parameters, extract them from the credentialData
+                    pendingCredential.certify(
+                        credentialData.toByteArray(),
+                        Clock.System.now(),
+                        Clock.System.now() + 30.days
+                    )
+                }
+            }
+        } catch (err: Throwable) {
+            documentStore.deleteDocument(document.identifier)
+            throw err
+        }
+        document.metadata.markAsProvisioned()
+        mutableState.emit(CredentialsIssued)
+        return document
+    }
+
+    /** Represents model's state */
+    sealed class State
+
+    /** Provisioning is not active */
+    data object Idle: State()
+
+    /** Provisioning is about to start */
+    data object Initial: State()
+
+    /** Connected to the provisioning server */
+    data object Connected: State()
+
+    /**
+     * Authorizing the user.
+     *
+     * When in this state, [provideAuthorizationResponse] must be called to authorize user using
+     * one of the methods in [authorizationChallenges]
+     *
+     * @param authorizationChallenges holds non-empty list of authorization methods and data
+     * necessary to use them.
+     */
+    data class Authorizing(
+        val authorizationChallenges: List<AuthorizationChallenge>
+    ): State()
+
+    /** Authorization response is being processed */
+    data object ProcessingAuthorization: State()
+
+    /** User was successfully authorized */
+    data object Authorized: State()
+
+    /** Credentials are being requested from the provisioning server */
+    data object RequestingCredentials: State()
+
+    /** Credentials are issued, provisioning has stopped */
+    data object CredentialsIssued: State()
+
+    /** Error occurred when provisioning, provisioning has stopped */
+    data class Error(
+        val err: Throwable
+    ): State()
```

Comment:
Do you think we should update `PresentmentModel` to also use `sealed class State` pattern?